File: 09_DataStructure_Algorithm/0518_day51/01_Permutations/Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> answer = new ArrayList<>();
        dfs(new ArrayList<>(), nums, answer);
        return answer;
    }

    private void dfs(List<Integer> cur, int[] nums, List<List<Integer>> answer) {
        if (cur.size() == nums.length) {
            answer.add(new ArrayList<>(cur));
            return;
        }

        for (int num : nums) {
            if (!cur.contains(num)) {
                cur.add(num);
                dfs(cur, nums, answer);
                cur.remove(cur.size() - 1);
            }
        }
    }
}

File: 09_DataStructure_Algorithm/0518_day51/01_Permutations/Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        answer = []

        n = len(nums)
        visited = [False] * n

        def permutations(cur):
            if len(cur) == n:
                answer.append(cur[:])
                return

            for idx in range(0, n):
                if not visited[idx]:
                    visited[idx] = True
                    cur.append(nums[idx])
                    permutations(cur)
                    visited[idx] = False
                    cur.pop()
        
        permutations([])

        return answer